### PR TITLE
Splines linear solver band

### PR DESCRIPTION
**Pull request by blegouix**

### CODE REVIEW

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+#include "splines_linear_problem.hpp"
+
+namespace ddc::detail {
+
+/**
+ * @brief A band linear problem dedicated to the computation of a spline approximation.
+ *
+ * The storage format is dense row-major. Lapack is used to perform every matrix and linear solver-related operations.
+ *
+ * @tparam ExecSpace The Kokkos::ExecutionSpace on which operations related to the matrix are supposed to be performed. Note: atm this is a placeholder for futur developments on GPU.
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:26]:
```suggestion
 * @tparam ExecSpace The Kokkos::ExecutionSpace on which operations related to the matrix are supposed to be performed.
```

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+class SplinesLinearProblemBand : public SplinesLinearProblem<ExecSpace>
+{
+public:
+    using typename SplinesLinearProblem<ExecSpace>::MultiRHS;
+    using SplinesLinearProblem<ExecSpace>::size;
+
+protected:
+    const std::size_t m_kl; // no. of subdiagonals
+    const std::size_t m_ku; // no. of superdiagonals
+    const std::size_t m_c; // no. of "rows" in q (in the sense of a band format storage)
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:38]:
```suggestion
    std::size_t m_kl; // no. of subdiagonals
    std::size_t m_ku; // no. of superdiagonals
    std::size_t m_c; // no. of "rows" in q (in the sense of a band format storage)
```

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+public:
+    /**
+     * @brief SplinesLinearProblemBand constructor.
+     *
+     * @param mat_size The size of one of the dimensions of the square matrix.
+     */
+    explicit SplinesLinearProblemBand(
+            const std::size_t mat_size,
+            const std::size_t kl,
+            const std::size_t ku)
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:51]:
```suggestion
            std::size_t const mat_size,
            std::size_t const kl,
            std::size_t const ku)
```

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        Kokkos::deep_copy(b_host, b);
+        int const info = LAPACKE_dgbtrs(
+                LAPACK_ROW_MAJOR,
+                transpose ? 'T' : 'N',
+                b_host.extent(0),
+                m_kl,
+                m_ku,
+                b_host.extent(1),
+                m_q.data(),
+                size(), // size() if LAPACK_ROW_MAJOR, m_c if LAPACK_COL_MAJOR
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:159]: Same question.

File: include/ddc/kernels/splines/splines_linear_problem_maker.hpp [hunk 5]
```diff
@@ -50,8 +50,11 @@ class SplinesLinearProblemMaker
             [[maybe_unused]] int const ku,
             [[maybe_unused]] bool const pds)
     {
-        // Placeholder while SplinesLinearProblemBand is not implemented
-        return std::make_unique<SplinesLinearProblemDense<ExecSpace>>(n);
+        if (2 * kl + 1 + ku >= n) {
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_maker.hpp:53]: Can you remove `[[maybe_unused]]` for kl and ku ?

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+         * while the column index is actually the column index of the matrix. Two layouts
+         * are supported by LAPACKE. The m_kl first lines are irrelevant for the storage of 
+         * the matrix itself but required for the storage of its LU factorization.
+         */
+        if (i >= std::max((std::ptrdiff_t)0, (std::ptrdiff_t)j - (std::ptrdiff_t)m_ku)
+            && i < std::min(size(), j + m_kl + 1)) {
+            // m_q(j * m_c + m_kl + m_ku + i - j) = aij; // LAPACK_COL_MAJOR
+            m_q((m_kl + m_ku + i - j) * size() + j) = aij; // LAPACK_ROW_MAJOR
+        } else {
+            assert(std::fabs(aij) < 1e-20);
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:111]: Could you provide an internal function that, given `(i, j)` from A_i,j, return a new couple on indexes representing the band and the position within this band ?

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    void setup_solver() override
+    {
+        int const info = LAPACKE_dgbtrf(
+                LAPACK_ROW_MAJOR,
+                size(),
+                size(),
+                m_kl,
+                m_ku,
+                m_q.data(),
+                size(), // size() if LAPACK_ROW_MAJOR, m_c if LAPACK_COL_MAJOR
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:129]: Can you provide a version using a 2D Kokkos View ? this parameter would then become m_q.stride(0) whatever the layout.

**Review by tpadioleau — CHANGES_REQUESTED** [review]
Almost good to me, just wondering if we could improve readability with layout and an intermediate function linked to the band structure

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp (hunk 7 above)
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_problem_band.hpp:129]: Yes that's true

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp (hunk 2 above)
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_problem_band.hpp:38]: Can you explain me why we do not need const here ?

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp (hunk 7 above)
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_problem_band.hpp]: Actually that's not, it has to be m_q.stride(0) if ROW_MAJOR and m_q.stride(1) if COL_MAJOR. But the 2D Kokkos::View makes sense for sure.

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        , m_kl(kl)
+        , m_ku(ku)
+        , m_c(2 * kl + ku + 1)
+        , m_ipiv("ipiv", mat_size)
+        /*
+         * The matrix itself stored in band format requires a (kl + ku + 1)*mat_size 
+         * allocation, but the LU-factorization requires an additional kl*mat_size block
+         */
+        , m_q("q", m_c, mat_size)
+    {
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:62]:
```suggestion
    Kokkos::View<int*, Kokkos::HostSpace> m_ipiv; // pivot indices
    Kokkos::View<double**, Kokkos::HostSpace> m_q; // band matrix representation

public:
    /**
     * @brief SplinesLinearProblemBand constructor.
     *
     * @param mat_size The size of one of the dimensions of the square matrix.
     */
    explicit SplinesLinearProblemBand(
            std::size_t const mat_size,
            std::size_t const kl,
            std::size_t const ku)
        : SplinesLinearProblem<ExecSpace>(mat_size)
        , m_kl(kl)
        , m_ku(ku)
        , m_ipiv("ipiv", mat_size)
        /*
         * The matrix itself stored in band format requires a (kl + ku + 1)*mat_size 
         * allocation, but the LU-factorization requires an additional kl*mat_size block
         */
        , m_q("q", 2 * kl + ku + 1, mat_size)
    {
```

Just removing m_c

**Pull request merged**